nearcore/src/state_sync.rs

```diff
@@ -237,432 +197,862 @@ async fn get_missing_part_ids_for_epoch(
+            CurrentDump::None => {}
+        };
+        match self.get_dump_state(&sync_header)? {
+            NewDump::Dump(mut dump, sender) => {
+                self.store_headers(&mut dump).await?;
+                self.start_upload_parts(sender, &dump).await;
+                self.new_dump(dump, *sync_header.hash())?;
+            }
+            NewDump::NoTrackedShards => {
+                self.current_dump = CurrentDump::Done(*sync_header.epoch_id());
+            }
+        };
+        Ok(())
+    }
 }
 
-/// return epoch_id and sync_hash of the latest complete epoch available locally.
-fn get_latest_epoch(
-    shard_id: &ShardId,
-    chain: &Chain,
+/// Main entry point into the state dumper. Initializes the state dumper and starts a loop that periodically
+/// checks whether there's a new epoch to dump state for.
+async fn state_sync_dump(
+    clock: Clock,
+    chain: Chain,
     epoch_manager: Arc<dyn EpochManagerAdapter>,
-) -> Result<Option<LatestEpochInfo>, Error> {
-    let head = chain.head()?;
-    tracing::debug!(target: "state_sync_dump", ?shard_id, "Check if a new complete epoch is available");
-    let hash = head.last_block_hash;
-    let header = chain.get_block_header(&hash)?;
-    let final_hash = header.last_final_block();
-    if final_hash == &CryptoHash::default() {
-        return Ok(None);
+    shard_tracker: ShardTracker,
+    runtime: Arc<dyn RuntimeAdapter>,
+    chain_id: String,
+    external: ExternalConnection,
+    mut iteration_delay: Duration,
+    validator: MutableValidatorSigner,
+    keep_running: Arc<AtomicBool>,
+    future_spawner: Arc<dyn FutureSpawner>,
+) -> anyhow::Result<()> {
+    tracing::info!(target: "state_sync_dump", "Running StateSyncDump loop");
+
+    // This is set to zero in some tests where the block production delay is very small (10 millis).
+    // In that case we'll actually just wait for 1 millisecond. The previous behavior was to call
+    // clock.sleep(ZERO), but setting it to 1 is probably fine, and works with the Instant below.
+    if iteration_delay == Duration::ZERO {
+        iteration_delay = Duration::milliseconds(1);
     }
-    let Some(sync_hash) = chain.get_sync_hash(final_hash)? else {
-        return Ok(None);
-    };
-    let final_block_header = chain.get_block_header(&final_hash)?;
-    let epoch_id = *final_block_header.epoch_id();
-    let epoch_info = epoch_manager.get_epoch_info(&epoch_id)?;
-    let epoch_height = epoch_info.epoch_height();
 
-    tracing::debug!(target: "state_sync_dump", ?final_hash, ?sync_hash, ?epoch_id, epoch_height, "get_latest_epoch");
+    let mut dumper = StateDumper::new(
+        clock.clone(),
+        chain_id,
+        validator,
+        shard_tracker,
+        chain,
+        epoch_manager,
+        runtime,
+        external,
+        future_spawner,
+    );
+    dumper.init(iteration_delay).await?;
+
+    let now = clock.now();
+    let mut check_head = Interval::new(now + iteration_delay, iteration_delay);
```

Comment:
maybe try this to remove the `mut` from  `mut iteration_delay: Duration`
```suggestion
    let min_iteration_delay = Duration::milliseconds(1);
    let mut check_head = Interval::new(now + iteration_delay, min_iteration_delay.max(iteration_delay));
```

Reply:
yeah good idea